.devops/overlay.nix

```diff
@@ -0,0 +1,20 @@
+final: prev:
+
+let
+  inherit (prev.stdenv) isAarch32 isAarch64 isDarwin;
+
+  darwinSpecific = if isAarch64 then {
+    inherit (prev.darwin.apple_sdk_11_0.frameworks) Accelerate MetalKit;
+  } else {
+    inherit (prev.darwin.apple_sdk.frameworks) Accelerate CoreGraphics CoreVideo;
+  };
+
+  osSpecific = if isDarwin then darwinSpecific else {};
+in
+
+{
+  llama-cpp = prev.callPackage ./package.nix osSpecific;
+  llama-cpp-opencl = prev.callPackage ./package.nix (osSpecific // { useOpenCL = true; });
+  llama-cpp-cuda = prev.callPackage ./package.nix (osSpecific // { useCuda = true; });
+  llama-cpp-rocm = prev.callPackage ./package.nix (osSpecific // { useRocm = true; });
```

Comment:
You almost never want to use `prev.callPackage` over `final.callPackage`. The heuristic is that you want the final versions of your dependencies.

And just in case that got lost, there's an alternative to multiple `callPackage`s, which is to use `.override`. Example: https://github.com/NixOS/nixpkgs/blob/0fe31382c0756f500b0cc6ddc7c1965afd28a983/pkgs/top-level/python-packages.nix#L14350-L14364.

Reply:
Thanks, this was really helpful. Updated.